### ISSUE TITLE
Define (and build) a rich tree model from a sort of manifest?

### ISSUE THREAD

**Issue by jimallman** [OpenTreeOfLife/phylostylotastic#18]
Following on our discussion of the main _pre-_visualization problem, i.e. How can we define and build a richer tree, with arbitrary properties on its elements? In one sense, this is a solved problem in that there are formats (e.g. NEXML) that can contain such a tree, but some big problems remain:
  - It's hard to build such a model, and hard to repeat that building w/o the scripts used.
  - Support for NEXML is a non-trivial feature that many viz problems will resist.
  - We lose touch with simpler, flatter source files (CSV, etc) that might be useful.

I'm pondering something like a Makefile, a sort of recipe that describes the rich tree we want, and how to build it from available artifacts. Something like this:
```json
{
    "--": "Example tree spec in JSON",
    "--": "(These lines are comments and will be ignored.)",
    "author": {"name":"Jim Allman", "email":"jim@ibang.com"},
    "data-sources": {
        "--": "These are arbitrarily named files/URLs referred to below",
        "main-tree": {"file":"./my_newick.tre"},
        "bamm": {"file":"./analyses/bamm.csv"},
        "edge-support": {"url":"https://example.com/support/2342334"}
    },
    "--": "Does this define a single-tree model? Or maybe more?",
    "tree": {
        "source": "main-tree",
        "publication": "http://dx.doi.org/10.12345/jim.example",
        "properties": {
            "--": "These named properties would be added to the output model."
            "--": "For visualization, they could be used in NexSS/TreeSS selectors.",
            "title": "Speciation of the North American honeybee"
        }
    },
    "nodes": {
        "properties": {
            "--": "These named properties would be added to each node in the output model."
            "--": "For visualization, they could be used in NexSS/TreeSS selectors.",
            "Wingspan": "TODO (see edge example below)",
        }
    },
    "edges": {
        "properties": {
            "--": "These named properties would be added to each node in the output model."
            "--": "For visualization, they could be used in NexSS/TreeSS selectors.",
            "Branch support": {
                "type":"integer",
                "source": {
                    "--": "Attach value from column 2 of CSV above, keyed by col 1 and node's 'id'",
                    "data": "edge-support",
                    "node-key":"id",
                    "source-key":{"column": 1},
                    "source-value": {"column": 2}
                }
            }
        }
    }
}
```
This is very rough, but hopefully suggestive. I'll add more later.

Naming possibilities: "TreeSpec"?

**Comment by hlapp** [comment]
Has anyone here looked at [`ggtree`](http://bioconductor.org/packages/devel/bioc/vignettes/ggtree/inst/doc/ggtree.html)? I know it's an R package, but if not "losing touch" with others is of any concern, and you want to explore the mental concept of building a tree viz from steps where a step adds on to what the previous ones have built, I would strongly recommend the [Grammar of Graphics](http://www.springer.com/gp/book/9780387245447), and as an example its implementation in R, [ggplot2](http://ggplot2.org). See also the `ggtree` paper in MEE: Yu G, Smith DK, Zhu H, Guan Y, Lam TT-Y. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. _Methods Ecol Evol._ 2016; [doi:10.1111/2041-210X.12628](http://doi.org/10.1111/2041-210X.12628)

Just as a word of caution, and of course that's entirely my $0.02. It is absolutely no challenge to come up with some new format for specifically your problem, including to make it so that it seems intuitive. And perhaps adoption by anyone else or interoperability is of no concern, and hence adding one more to the gazillion formats and mechanisms for tree viz is just fine too. However, I would urge Open Tree to try to look beyond that.

**Comment by sckott** [comment]
I agree that perhaps exploring whether grammar of graphics can support all phylogenetic tree viz concepts seems worth doing. There's a python port of gggplot2 (which ggtree builds on),  https://github.com/yhat/ggplot, but maybe http://www.dendropy.org/ is the canonical tree plotting tool in python?

**Comment by jimallman** [comment]
Thanks @hlapp and @sckott, I'll take a look at these! We did discuss `ggtree` in today's call. 

> And perhaps adoption by anyone else or interoperability is of no concern, and hence adding one more to the gazillion formats and mechanisms for tree viz is just fine too. 

Oh no, interoperability is _exactly_ what we're after. If there's an existing solution, I don't want to re-invent the wheel. But this has been identified as a pain point, so it seems any existing solutions aren't widely known, or are so specific to a particular set of tools as to limit interop.

**Comment by hlapp** [comment]
> If there's an existing solution, I don't want to re-invent the wheel. But this has been identified as a pain point, so it seems any existing solutions aren't widely known, or are so specific to a particular set of tools as to limit interop.

That's jumping to a conclusion, no? In other words, there may be many reasons why an approach is not widely used yet. That it is too specific and not interoperable is only one of them. That it is new and in proof-of-concept stage is another one, and there are still more.

And if an approach isn't exactly yet what you need, that doesn't mean either it can't be built upon?

Again, if interoperability is what you are after, I would treat creating a new format as the very last resort, because it's really one of the easiest things to do (which is why so many do exactly that). And the vast majority of new formats do not get adopted by anyone, despite the occasional success stories.

**Comment by arlin** [comment]
Hilmar, you should join our discussions weekly (Tuesdays, noon Eastern).  A new format isn't necessarily what we need, maybe just a combination of existing standards.  Imagine this.  Put the tree in Newick.  No one will have to use anything other than Newick (after all, why not just accept the past 25 years?).  From here, we need a few things.  We need a language for specifying parts of the tree-- like phyloreferencing.   And we need a language to specify graphical elements.  For instance, "bracket clade (A, B) and label it 'Poriphera' in red Helvetica 15" (maybe too detailed).  We need a way to refer to data associated with the tree or with nodes of the tree, but perhaps this could be very generic, based on a model of tables.

**Comment by jimallman** [comment]
To be clear, this issue is about finding (or inventing) a declarative, _language-agnostic_ description of how to compose a rich tree model from multiple, simple sources. This ignores the later step of plotting and styling a finished graph, though these are often addressed in the same tool.

At @hlapp 's prompting, I'm searching for existing standards and finding... not much so far. There are of course many existing viz tools, generally based in Wilkinson's "grammar of graphics", but typically the user composes a visualization directly in R, Python, or some other language. We're looking for something that would allow users of **different languages and tools** to easily share rich models without resorting to a monolithic format like NeXML.

Wilkinson himself went on to develop [GPL (Graphics Programming Language)](https://www.ibm.com/support/knowledgecenter/SSLVMB_20.0.0/com.ibm.spss.statistics.help/gpl_intro_basics.htm) (non-trivial example [here](https://andrewpwheeler.wordpress.com/2016/08/12/plotting-panel-data-with-many-lines-in-spss/)) for IBM's [SPSS](http://www.ibm.com/analytics/us/en/technology/spss/), but GPL (since renamed nViZn) seems to be a closed and proprietary standard. IBM's team also developed a JSON serialization called [VizJSON](http://www.ibm.com/developerworks/library/bd-vizjson/), but it's also proprietary with no public documentation that I can find. Tableau also uses a proprietary [XML format `.tds`](http://www.theinformationlab.co.uk/2013/12/02/tableau-file-types-and-extensions/) (more details [here](https://www.interworks.com/blog/ksloan/2015/04/09/tableau-performance-checklist-data-use-one-tds-file)) to capture data source information, metadata, calculated fields, etc. for sharing and re-use.

Is there an open source, language-agnostic alternative? So far, the most promising candidate has been right under my nose--the [declarative Vega serialization](https://github.com/vega/vega/wiki/Documentation) that we use in [Tree Illustrator](https://github.com/opentreeoflife/tree-illustrator). Vega is yet another implementation of the grammar of graphics, but one that uses JSON to weave data, style, and [d3 layouts](https://github.com/d3/d3/blob/master/API.md#hierarchies-d3-hierarchy) into a finished graphic. We might use a subset of their specification (see esp. [Data](https://github.com/vega/vega/wiki/Data) and [Data Transforms](https://github.com/vega/vega/wiki/Data-Transforms)) to describe how to build the data model, since any existing viz tool should be able to add a JSON parser without much trouble. Other suggestions are of course welcome!

**Comment by jimallman** [comment]
> We're looking for something that would allow users of different languages and tools to easily share rich models without resorting to a monolithic format like NeXML.

Of course, one could argue that adding full NeXML support to more tools is the smarter path.

**Comment by jimallman** [comment]
@rhr suggests another approach, aiming for **TSTTCPW** (ie, the simplest thing that could possibly work). We'd outline a simple, uniform process for composing a rich tree model from the most commonly used file formats (Newick for trees, CSV for metadata). This would make lots of assumptions, but would also lighten the burden for apps trying to use the standard, while putting more responsibility on the party sharing the data.

The basic idea is a flat collection of files whose "numbered" filenames describe a composition sequence, for example:
```bash
01-american-butterflies.tre      # tree as Newick (or Extended Newick)
02-taxon-names.csv               # mapped to node ids, ordinal positions (ape convention)
04-habitats.csv
05-anatomic-traits.csv
06-support-values.csv
07-taxon-images.csv              # PhyloPic URLs, for selected taxa?
```
This would guide the consuming program to construct the tree in `american-butterflies.tre`, then parse each of the remaining CSV files in turn, adding new properties and values to the tree elements [in the manner described here](https://www.r-phylo.org/wiki/HowTo/InputtingData). The (optional?) numbered steps would allow building on previously added properties, esp. taxon names.

Note that we make no assumptions about the types or classes used in the consuming program; we simply assume that it's capable of constructing a tree and assigning properties to its basic elements (tree, node, edge).

Complications like multiple trees, shared metadata, data-type coercion, etc. could be handled fairly easily, but each would complicate the standard. At some point we'd need the sort of structured configuration/manifest file described above. But would this simple approach suffice for common uses? Is this TSTTCPW?

(@rhr, I've simplified our discussion even further while writing the above description. Please step in if I've gone too far or mis-communicated your approach here.)

**Comment by daisieh** [comment]
So is the idea that the flat collection would represent the same contents as a fully annotated Nexml or Nexson file? It seems like it's very similar, except with 1) a reliance on Newick for tree encoding, vs the nodes-and-edges model of Nexml/Nexson, and 2) un-interleaving the annotations from the nodes and edges and putting them into separate csv files instead. Am I misunderstanding this?

**Comment by nfranz** [comment]
I am just reading along, but it seems to me that it will be hard to
separate the more pragmatic solution bits from this underlying theme? To
what extent should the "rich tree model" be a vehicle to achieve a very
information rich but nevertheless mostly singular/synthetic perspective; or
be also a vehicle to represent and mediate incremental updates and
conflicting views? "The simplest thing that could possibly work", to me is
not specific enough in addressing "work for what and whom, exactly?" And by
leaving that too open, perhaps agreements take longer to reach than they
need to.

Best, Nico

**Comment by jimallman** [comment]
> So is the idea that the flat collection would represent the same contents as a fully annotated Nexml or Nexson file? It seems like it's very similar, except with 1) a reliance on Newick for tree encoding, vs the nodes-and-edges model of Nexml/Nexson, and 2) un-interleaving the annotations from the nodes and edges and putting them into separate csv files instead. Am I misunderstanding this?

Yes, on all counts. It's primarily a concession to:
- the apparent difficulty of fully supporting Nexml/Nexson, and
- the common use of these "flat" formats vs. more complex ones

(I'm relying on @jhcepas and @rhr, respectively, who have strongly stated these points.)

**Comment by jimallman** [comment]
> "The simplest thing that could possibly work", to me is not specific enough in addressing "work for what and whom, exactly?"

Well said. I'm looking at this as a software guy, not a scientist, so I'm hoping to elicit just this kind of feedback from those who are closer to the work.

**Comment by arlin** [comment]
@nfranz we're having a videochat in half an hour, at noon EST.  you are welcome to join.  just email me and i'll send you the hangout link.

**Comment by nfranz** [comment]
Hi Arlin:

   Thanks. I might be able to join for a while. So yes please.

Best, Nico

**Comment by arlin** [comment]
ok, we will be at http://bit.ly/1GWSrDf (google+)
